src/compo/mls_o3_nc2ioda.py

```diff
@@ -0,0 +1,381 @@
+        for v in varsToAddUnits:
+            if(v != 'valKey' and v != 'errKey'):
+                vkey = (v, 'MetaData')
+                if('pressure' in v.lower()):
+                    self.varAttrs[vkey]['units'] = 'Pa'
+                elif(v == 'dateTime'):
+                    self.varAttrs[vkey]['units'] = 'seconds since 1993-01-01T00:00:00Z'
+                elif('angle' in v.lower() or 'latitude' in v.lower() or 'longitude' in v.lower()):
+                    self.varAttrs[vkey]['units'] = 'degrees'
+                elif('flag' in v.lower()):
+                    self.varAttrs[vkey]['units'] = 'unitless'
+                elif('prior' in v.lower()):
+                    self.varAttrs[vkey]['units'] = 'ppmv'
+                else:
+                    self.varAttrs[vkey]['units'] = 'unitless'
+
+    # Read data needed from raw MLS file.
+    def _read_nc(self, filename, ifile, maxfile):
+        print("Reading: {}".format(filename))
+        ncd = nc.Dataset(filename, 'r')
+        end_of_file = len(ncd[ioda2nc['latitude']][:])
+        # unles nrt, use all profiles in file.
+        start = 0
+        end = end_of_file
+        # if it is nrt, and the last file use all but first two and last 3 profiles
+        if(ifile == maxfile and self.nrt):
+            print("last file:{}".format(filename))
+            start = 2
+            end = end_of_file - 3
+        # otherwise if nrt skip first 2 and last 8 profiles to avoid duplicates.
+        elif(self.nrt):
+            start = 2
+            end = end_of_file - 8
+
+        d = {}
+        for k in list(ioda2nc.keys()):
+            if (k == 'air_pressure'):
+                d[k] = ncd[ioda2nc[k]][...]*100.  # convert to Pa
+                d[k].mask = False
+            else:
+                d[k] = ncd[ioda2nc[k]][start:end, ...]
+                d[k].mask = False
+
+            if(k == 'valKey' or k == 'precision'):
+                d[k] = d[k]*1e6  # convert mol/mol to PPMV
+        return d
+
+    def _calc_error(self, o3, o3_prec, lev):
+        # Observation error estimates from MLS.
+        oe = ['na', 'na', 'na', 'na', 'na', 'na', 'na', 0.02, 0.02, 0.02, 0.02, 0.035, 0.05, 0.05, 0.05,
+              0.125, 0.2, 0.2, 0.2, 0.2, 0.2, 0.225, 0.25, 0.275,
+              0.3, 0.3, 0.3, 0.3, 0.3, 0.275, 0.25, 0.225, 0.2, 0.2,
+              0.2, 0.2, 0.2, 0.15, 0.1, 0.1, 0.1, 0.15, 0.2, 0.2, 0.2,
+              0.3, 0.3, 0.3, 0.3]
+        ooe = oe[lev]
+        # inflate errors for specific levels. (note zero based index)
+        if (lev == 7):  # 261 hPa
+            ooe = ooe + (0.30 * abs(o3))  # 0.3 is my conservative guess
+        elif (lev == 8):  # 215 hPa
+            ooe = ooe + (0.20 * abs(o3))
+        elif (lev == 9):  # 177 hPa
+            ooe = ooe + (0.125 * abs(o3))
+        elif (lev == 10 or lev == 11 or lev == 12):  # 150-100 hPa
+            ooe = ooe + (0.05 * abs(o3))
+
+        ooe = np.sqrt(max((0.5*ooe)**2+(o3_prec)**2, 1.e-6))
+        return ooe
```

Comment:
This error specification is enabled for MLS when qcOn. _calc_error is called in _do_qc. Should those ad-hoc errors specifications decoupled from the qc filtering? Meaning we could have a separate option for replacing the errors?

Reply:
I mean, I guess? But the two options I was thinking were do the QC and error calculation downstream or upstream. I don't know that we're that motivated to push more QC/error calculation within the ufo, so I'm dubious we'd use anything but the default.

Reply:
To clarify I would see it more flexible to have the option to do the pre qc filtering and not the ad-hoc errors and conversely. For that _calc_error should be called outside of _do_qc within a condition triggered by an option. It is more a suggestion than a requirement here. I am not completely new to MLS but it's been more than 10 years now... I remember to have to correct those errors for DA.

Reply:
Ok, I broke it out, and moved it out of _do_qc.